kernprof.py

```diff
@@ -155,53 +155,46 @@ def find_script(script_name):
-    parser.allow_interspersed_args = False
-    parser.add_option('-l', '--line-by-line', action='store_true',
-        help="Use the line-by-line profiler from the line_profiler module "
-        "instead of Profile. Implies --builtin.")
-    parser.add_option('-b', '--builtin', action='store_true',
-        help="Put 'profile' in the builtins. Use 'profile.enable()' and "
-            "'profile.disable()' in your code to turn it on and off, or "
-            "'@profile' to decorate a single function, or 'with profile:' "
-            "to profile a single section of code.")
-    parser.add_option('-o', '--outfile', default=None,
-        help="Save stats to <outfile>")
-    parser.add_option('-s', '--setup', default=None,
+def main():
+    def positive_float(value):
+        val = float(value)
+        if val <= 0:
+            raise ArgumentError
+        return val
+
+    parser = ArgumentParser()
+    parser.add_argument('-V', '--version', action='version', version=__version__)
+    parser.add_argument('-l', '--line-by-line', action='store_true',
+        help="Use the line-by-line profiler instead of cProfile. Implies --builtin.")
+    parser.add_argument('-b', '--builtin', action='store_true',
+        help="Put 'profile' in the builtins. Use 'profile.enable()'/'.disable()', "
+            "'@profile' to decorate functions, or 'with profile:' to profile a "
+            "section of code.")
+    parser.add_argument('-o', '--outfile',
+        help="Save stats to <outfile> (default: 'scriptname.lprof' with "
+            "--line-by-line, 'scriptname.prof' without)")
+    parser.add_argument('-s', '--setup',
         help="Code to execute before the code to profile")
-    parser.add_option('-v', '--view', action='store_true',
-        help="View the results of the profile in addition to saving it.")
-    parser.add_option('-u', '--unit', default='1e-6', type=float,
-        action='callback', callback=strictly_positive,
-        help="Output unit (in seconds) in which the timing info is to be "
-            "displayed (for --view). Defaults to 1e-6.")
-    parser.add_option('--skip-zero', action='store_true',
-        help="Hide functions which have not been called (for --view).")
+    parser.add_argument('-v', '--view', action='store_true',
+        help="View the results of the profile in addition to saving it")
+    parser.add_argument('-u', '--unit', default='1e-6', type=positive_float,
+        
+        help="Output unit (in seconds) in which the timing info is "
+        "displayed (default: 1e-6)")
+    parser.add_argument('-z', '--skip-zero', action='store_true',
+        help="Hide functions which have not been called")
 
-    if not sys.argv[1:]:
-        parser.print_usage()
-        sys.exit(2)
+    parser.add_argument('script', help="The python script file to run")
+    parser.add_argument('args', nargs='...', help="Optional script arguments")
 
-    options, args = parser.parse_args()
+    options = parser.parse_args()
```

Comment:
You can pass in `args` to `parser.parse_args`. Could you re-add the `args` argument to `main`, and use `sys.exit(main(sys.argv))` instead of `main()` to preserve the existing signature?

Reply:
I can but why? I never saw any script do that as `sys.argv` is the default of `parse_args`.

Reply:
The main reason is that I don't want to change the signature of the main method.

A secondary reason is that I always prefer to rely on arguments and passing locals rather than globals. There is a reason argparse allows you to specify `argv` rather than always assuming you are using the global `sys.argv`. 

Although I haven't done it with this library, I often use parameterized `argv` that overrides `sys.argv`. This is especially useful for unit tests.

Reply:
Ok I changed that, but note that the old version completely ignored the `args` parameter and just overwrote it with the output of `parser.parse_args()` (without arguments) so it in fact used the global `sys.argv`.  (I think that what I wanted to remove the argument.)